### PR TITLE
[FLINK-7784] [kafka-producer] Don't fail TwoPhaseCommitSinkFunction when failing to commit during job recovery

### PR DESCRIPTION
**Pull request by GJL**
## What is the purpose of the change
This makes it possible to configure the TwoPhaseCommitSinkFunction's behaviour w.r.t. transaction timeouts.

## Brief change log
  - *Introduce transaction timeouts to TwoPhaseCommitSinkFunction.* 
  - *Timeout can be used to generate warnings if the transaction's age approaches the timeout.*
  - *If an exception is thrown during job recovery, the sink can be configured not to propagate the exception and instead log it on ERROR level.*

## Verifying this change
This change added tests and can be verified as follows:
  - *Extended unit tests for TwoPhaseCommitSinkFunction to test added functionality*
  - *Manually verified the change by running a job with a FlinkKafka011Producer with checkpoint interval 27000 and transaction.timeout.ms = 30000. Warnings were generated correctly.*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (**yes** / no)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (**yes** / no)
  - If yes, how is the feature documented? (not applicable / docs / **JavaDocs** / not documented)

### CODE REVIEW

**Comment by GJL** [comment]
@aljoscha

**Comment by aljoscha** [comment]
The changes look very good! @pnowojski could you please also have a look at this?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunction.java [hunk 1]
```diff
@@ -58,18 +61,37 @@
...
+	 * If true, any exception thrown in {@link #recoverAndCommit(Object)} will be caught instead of
+	 * propagated.
+	 */
+	private boolean failureOnCommitAfterTransactionTimeoutDisabled;
+
+	/**
+	 * If a transaction's elapsed time reaches this percentage of the transactionTimeout, a warning
+	 * message will be logged. Value must be in range [0,1]. Negative value disables warnings.
+	 */
+	private double transactionTimeoutWarningRatio = -1;
```
Comment (pnowojski) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunction.java]: I would prefer to implement this sanity check in `FlinkKafkaProducer011`, since it's a walk around Kafka's bug and unlikely to be useful in general case. And after Kafka `0.11.0.2` or `1.0.0` release we won't need this code anymore.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunction.java [hunk 2]
```diff
@@ -293,26 +330,110 @@ public void initializeState(FunctionInitializationContext context) throws Except
...
-		LOG.debug("{} - started new transaction '{}'", name(), currentTransaction);
+		currentTransaction = beginTransaction0();
+		log.debug("{} - started new transaction '{}'", name(), currentTransaction);
+	}
+
+	/**
+	 * This method must be the only place to call {@link #beginTransaction()} to ensure that the
+	 * {@link TransactionHolder} is created at the same time.
+	 */
+	private TransactionHolder<TXN> beginTransaction0() throws Exception {
```
Comment (pnowojski) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunction.java]: this `0` in method's name doesn't help in anything. Please rename it to either `beginTransactionHolder`, `beginTransactionWrapper`, `beginTransactionInternal`, `beginTransactionAndStartTimeoutTimer` or `beginTransactionAndMarkTime`

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunction.java [hunk 3]
```diff
@@ -58,18 +61,37 @@
...
 
-	@Nullable
-	protected TXN currentTransaction;
 	protected Optional<CONTEXT> userContext;
 
 	protected ListState<State<TXN, CONTEXT>> state;
 
+	private final ListStateDescriptor<State<TXN, CONTEXT>> stateDescriptor;
+
+	private TransactionHolder<TXN> currentTransaction;
```
Comment (pnowojski) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunction.java]: please rename this field, it's clashing with `currentTransaction()` method, which is confusing, especially that they have different types.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunction.java [hunk 4]
```diff
@@ -293,26 +330,110 @@ public void initializeState(FunctionInitializationContext context) throws Except
...
+	private TransactionHolder<TXN> beginTransaction0() throws Exception {
+		return new TransactionHolder<>(beginTransaction(), clock.millis());
+	}
+
+	/**
+	 * This method must be the only place to call {@link #recoverAndCommit(Object)} to ensure that
+	 * the configuration parameters {@link #transactionTimeout} and
+	 * {@link #failureOnCommitAfterTransactionTimeoutDisabled} are respected.
+	 */
+	private void recoverAndCommit(TransactionHolder<TXN> transactionHolder) {
```
Comment (pnowojski) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunction.java]: ditto: overloading adds confusion, because it suggests that both methods (`recoverAndCommit(TXN)` and `recoverAndCommit(TransactionHolder)`) are equally valid and could be used interchangeably. 

As above, rename to `recoverAndCommitHolder`, `recoverAndCommitWrapper`, `recoverAndCommitInternal`, `recoverCommitAndHandleTimeout`

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunction.java [hunk 5]
```diff
@@ -293,26 +330,110 @@ public void initializeState(FunctionInitializationContext context) throws Except
...
+		try {
+			logWarningIfTimeoutAlmostReached(transactionHolder);
+			recoverAndCommit(transactionHolder.handle);
+		} catch (final Exception e) {
+			final long elapsedTime = clock.millis() - transactionHolder.transactionStartTime;
+			if (failureOnCommitAfterTransactionTimeoutDisabled && elapsedTime > transactionTimeout) {
+				log.error("Error while committing transaction {}. " +
```
Comment (pnowojski) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunction.java]: `"Error while committing transaction {}. Data loss might occurred"`

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunction.java [hunk 6]
```diff
@@ -362,4 +483,39 @@ public void setContext(Optional<CONTEXT> context) {
 			this.context = context;
 		}
 	}
+
+	/**
+	 * Adds metadata (currently only the start time of the transaction) to the transaction object.
+	 */
+	@VisibleForTesting
+	static class TransactionHolder<TXN> {
```
Comment (pnowojski) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunction.java]: `TransactionWrapper`?

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunctionTest.java [hunk 7]
```diff
@@ -194,29 +348,4 @@ public String toString() {
 			return String.format("FileTransaction[%s]", tmpFile.getName());
 		}
 	}
-
-	private static class TestContext implements AutoCloseable {
```
Comment (pnowojski) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunctionTest.java]: Why did you remove this encapsulation?

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunctionTest.java [hunk 8]
```diff
@@ -35,60 +42,101 @@
...
+
+	private File targetDirectory;
+
+	private File tmpDirectory;
+
+	@Mock
+	private Clock mockClock;
+
+	@Mock
+	private Logger mockLogger;
```
Comment (pnowojski) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunctionTest.java]: As above. Will all of those `mockLogger` tests fail, if you replace:
```
			log.warn("Transaction {} has been open for {} ms. " +
					"This is close to or even exceeding the transaction timeout of {} ms.",
				transactionHolder.handle,
				elapsedTime,
				transactionTimeout);
```
with 
```
			log.warn("Transaction {} has been open for too long. " +
					"This is close to or even exceeding the transaction timeout of {} ms.",
				transactionHolder.handle,
				transactionTimeout);
```
or
```
			log.warn(String.format("Transaction {} has been open for {} ms. " +
					"This is close to or even exceeding the transaction timeout of {} ms.",
				transactionHolder.handle,
				elapsedTime,
				transactionTimeout));
```
?
If so, then this is a perfect example my I consider mockito to be the definition of evil. Using mockito in tests in 99% cases is duplicating/copying productional code into tests, basically repeating the implementation, which is super fragile and brakes on even the tiniest refactors. While good tests instead should check/assert the actual effects - not whether specific method calls were called and how many times they were called.

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunctionTest.java [hunk 9]
```diff
@@ -35,60 +42,101 @@
...
+	private OneInputStreamOperatorTestHarness<String, Object> harness;
+
+	private AtomicBoolean throwException = new AtomicBoolean();
+
+	private File targetDirectory;
+
+	private File tmpDirectory;
+
+	@Mock
+	private Clock mockClock;
```
Comment (pnowojski) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunctionTest.java]: please do not use mockito for classes that are so easy to implement and mock in old fashion way (creating `SettableClock` seems to be super easy).

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunctionTest.java (hunk 8 above)
Comment (GJL) [reply flink-streaming-java/src/test/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunctionTest.java]: The test is asserting on the presence of the substring `This is close to or even exceeding the transaction timeout` in the log message. All your changes to the code would still pass the test except for the 2nd case because there is an assert on the elapsed time. Mockito is used here so that the argument passed to `.warn` can be captured. Imo this is not evil as no behavior is mocked, and actual effects are tested.

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunctionTest.java (hunk 7 above)
Comment (GJL) [reply flink-streaming-java/src/test/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunctionTest.java]: I am using 
```
	@Rule
	public TemporaryFolder folder = new TemporaryFolder();
```
to create test folders are files which are cleaned up automatically. Since the rule must be a public field in the test, it would have required more work to keep the `TestContext`.

File: flink-connectors/flink-connector-kafka-0.11/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer011.java [hunk 10]
```diff
@@ -442,13 +445,31 @@ public FlinkKafkaProducer011(
...
 			checkState(timeout < Integer.MAX_VALUE && timeout > 0, "timeout does not fit into 32 bit integer");
 			this.producerConfig.put(ProducerConfig.TRANSACTION_TIMEOUT_CONFIG, (int) timeout);
 			LOG.warn("Property [%s] not specified. Setting it to %s", ProducerConfig.TRANSACTION_TIMEOUT_CONFIG, DEFAULT_KAFKA_TRANSACTION_TIMEOUT);
 		}
 
+		// Enable transactionTimeoutWarnings to avoid silent data loss
+		// See KAFKA-6119 (affects versions 0.11.0.0 and 0.11.0.1):
+		// The KafkaProducer may not throw an exception if the transaction failed to commit
+		if (semantic == Semantic.EXACTLY_ONCE) {
+			final Object object = this.producerConfig.get(ProducerConfig.TRANSACTION_TIMEOUT_CONFIG);
```
[the remark this answers is not in the document]
Comment (GJL) [reply flink-connectors/flink-connector-kafka-0.11/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer011.java]: Couldn't find any.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunction.java (hunk 3 above)
Comment (GJL) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunction.java]: Ok will do.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunction.java [hunk 11]
```diff
@@ -58,18 +61,37 @@
...
+	/**
+	 * Specifies the maximum time a transaction should remain open.
+	 */
+	private long transactionTimeout = Long.MAX_VALUE;
+
+	/**
+	 * If true, any exception thrown in {@link #recoverAndCommit(Object)} will be caught instead of
+	 * propagated.
+	 */
+	private boolean failureOnCommitAfterTransactionTimeoutDisabled;
```
[the remark this answers is not in the document]
Comment (GJL) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunction.java]: Ok, will do.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunction.java (hunk 4 above)
Comment (GJL) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunction.java]: Will rename.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunction.java (hunk 5 above)
Comment (GJL) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunction.java]: Good. Will add a statement to the log message.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunction.java (hunk 2 above)
Comment (GJL) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunction.java]: Will rename.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunction.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (GJL) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunction.java]: Renamed to `failurePropagationAfterTransactionTimeoutDisabled` as this is closer to the method that controls this field.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunction.java (hunk 1 above)
Comment (GJL) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunction.java]: If exceeding the `transactionTimeout` can result in data loss, I prefer a flag here (i.e., `TwoPhaseCommitSinkFunction`). You may want to be warned before something bad happens so that parameters can be tuned.

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunctionTest.java (hunk 9 above)
Comment (GJL) [reply flink-streaming-java/src/test/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunctionTest.java]: I agree with you that overusing mocks is an anti-pattern. Changed it to a real implementation. Please have a look. I am not sure if this is better now because after all, the only mocked function on Clock was `.millis()`.

File: flink-connectors/flink-connector-kafka-0.11/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer011.java [hunk 12]
```diff
@@ -922,6 +945,22 @@ private void readObject(java.io.ObjectInputStream in) throws IOException, ClassN
...
+	 * transactions during recovery of the job. If a Kafka transaction is timed out, a commit will
+	 * never be successful. Hence, use this feature to avoid recovery loops of the Job. Exceptions
+	 * will still be logged to inform the user that data loss might have occurred.
+	 *
+	 * <p>Note that we use {@link System#currentTimeMillis()} to track the age of a transaction.
+	 * Moreover, only exceptions thrown during the recovery are caught, i.e., the producer will
+	 * attempt at least one commit of the transaction before giving up.</p>
+	 */
+	@Override
+	public FlinkKafkaProducer011<IN> disableFailurePropagationAfterTransactionTimeout() {
```
[the remark this answers is not in the document]
Comment (GJL) [reply flink-connectors/flink-connector-kafka-0.11/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer011.java]: Thanks. Moved it.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunction.java (hunk 6 above)
Comment (GJL) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunction.java]: I will leave it as is. It's an internal class and it is clear what it does.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunction.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (pnowojski) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunction.java]: Part of my concern was that the name is quite long (same applies for the methods). Maybe `ignoreFailuresAfterTimeout`? Or `ignoreFailuresAfterTransactionTimeout`?

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunctionTest.java (hunk 8 above)
Comment (pnowojski) [reply flink-streaming-java/src/test/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunctionTest.java]: Are you sure? `Logger.warn(String, Object...)`, `Logger.warn(String, Object)`, `Logger.warn(String, Object, Object)` and `Logger.warn(String)` are different methods.

This shows, that you are not testing for the effects (warning message being logged somewhere), but you test whether one particular method was called or not.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunction.java (hunk 1 above)
Comment (pnowojski) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunction.java]: I'm not convinced, since this will make our public api larger and more difficult to maintain. However if @aljoscha is ok with that I will yield ;)

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunctionTest.java (hunk 8 above)
Comment (GJL) [reply flink-streaming-java/src/test/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunctionTest.java]: True, you are right. One would need to implement the `org.slf4j.Logger` interface.
- http://projects.lidalia.org.uk/slf4j-test/ provides a `org.slf4j.Logger` implementation for unit testing but this approach comes with other problems (e.g. https://github.com/Mahoney/slf4j-test/issues/15, [NOP if log level is  disabled](https://github.com/Mahoney/slf4j-test/blob/master/src/main/java/uk/org/lidalia/slf4jtest/TestLogger.java#L449)

Right now I don't see a way to do it properly. Any help is welcome.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunction.java (hunk 1 above)
Comment (aljoscha) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunction.java]: I tend to agree with @pnowojski about the API surface but I think in this case this is a valid safety net for possible future transaction sinks.

File: flink-connectors/flink-connector-kafka-0.11/src/test/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer011Tests.java [hunk 13]
```diff
@@ -83,49 +79,6 @@ public void before() {
 		extraProperties.put("isolation.level", "read_committed");
 	}
 
-	@Test(timeout = 30000L)
```
[the remark this answers is not in the document]
Comment (GJL) [reply flink-connectors/flink-connector-kafka-0.11/src/test/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer011Tests.java]: The tests that I removed were already in `FlinkKafkaProducerTests`. Probably some copy paste error.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunction.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (GJL) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunction.java]: Ok, I like `ignoreFailuresAfterTransactionTimeout`

[the remark this answers is not in the document]
Comment (GJL) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunction.java]: Ok, I like `ignoreFailuresAfterTransactionTimeout`. The method name is now the same, though.

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunctionTest.java (hunk 8 above)
Comment (GJL) [reply flink-streaming-java/src/test/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunctionTest.java]: The test does not rely on any mocks anymore. I am not 100% happy with it because we use log4j `1.x` which is not maintained anymore and in log4j `2.x`, the APIs have changed a lot: http://logging.apache.org/log4j/2.x/manual/customconfig.html#AddingToCurrent

File: flink-connectors/flink-connector-kafka-0.11/src/test/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer011Tests.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (pnowojski) [reply flink-connectors/flink-connector-kafka-0.11/src/test/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer011Tests.java]: Yes that's right. Code of this test is also in `FlinkKafkaProducerTest`.

**Comment by tzulitai** [comment]
LGTM, thanks for the work @GJL. Merging this ..

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunction.java [hunk 14]
```diff
@@ -312,20 +347,104 @@ public void initializeState(FunctionInitializationContext context) throws Except
...
+	/**
+	 * This method must be the only place to call {@link #recoverAndCommit(Object)} to ensure that
+	 * the configuration parameters {@link #transactionTimeout} and
+	 * {@link #ignoreFailuresAfterTransactionTimeout} are respected.
+	 */
+	private void recoverAndCommitInternal(TransactionHolder<TXN> transactionHolder) {
+		try {
+			logWarningIfTimeoutAlmostReached(transactionHolder);
+			recoverAndCommit(transactionHolder.handle);
+		} catch (final Exception e) {
```
Comment (XComp) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunction.java:371]: @GJL @pnowojski I came across this method today and was initially confused that this is actually possible, i.e. catching and then possibly re-throwing a checked `Exception` without mentioning it in the method signature. But apparently, this is valid Java as a special case of `Exception` handling where the checked `Exception` is actually a `RuntimeException`.

That said, this `catch` block should actually focus on `RuntimeExceptions` because `logWarningIfTimeoutAlmostReached` and `recoverAndCommit` both are not considering checked `Exceptions`. I'm just asking because I got curious whether I miss something here and this `catch` was done intentionally or whether this minor thing just slipped through...

Comment (GJL) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunction.java:371]: I think changing `Exception` to `RuntimeException` won't be semantically equivalent due to the possibility of [sneaky throws](https://www.baeldung.com/java-sneaky-throws#sneaky). My memory about this code is a bit fuzzy but I assume there can be sneaky throws from within `recoverAndCommit` because the `commit` callback is public API.

Comment (XComp) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunction.java:371]: Cool thanks for clarification. That makes sense. And thanks for the pointer to the sneaky throw example. 👍
